added pause/resume to Tracker.cs

martinmmueller opened the pull request:
# Motivation for the Pull request
I wanted the option to pause and resume trackers without dumping the data that was already collected (**StartRecording** creates a new **DataTable**), so I quickly built in a **Pause** and a **Resume** function with an enum-based state to keep track of what the tracker is currently doing.

Maybe this could be useful to others as well, so I thought I'd put the edit up here.

# Some notes:
* I realise the state enum I added makes the 'recording' field kind of redundant but I left that in place to not break any possible dependencies.
* I added some debug warnings for telling the tracker to start/pause/stop etc. when it does not really make sense. Ideally the tracker would only react to commands when it is in the matching state in the first place, but I did not want to change to much about the base class, so I left it at warnings for now. If those don't match your policy for logging/ error handling they could be escalated to full Exceptions (like you do in **RecordRow**), or just removed so the Tracker just silently does its thing.

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -44,7 +45,11 @@ public string DataName
         public bool Recording { get { return recording; } }
```
We can make this property return `currentState == TrackerState.On` and remove the `recording` field completely. Calls to `recording` in this class should be replaced with `currentState == TrackerState.On`

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -87,17 +92,58 @@ public void RecordRow()
         {
+            if (currentState == TrackerState.On)
+            {
+                Debug.LogWarning($"Start command received for tracker in state: '{TrackerState.On}'." +
```
Use `Utilities.UXFDebugLogWarning` instead

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -87,17 +92,58 @@ public void RecordRow()
         {
+            if (currentState != TrackerState.On)
+            {
+                Debug.LogWarning($"Stop command received for tracker in state: '{currentState}'." +
```
Use `Utilities.UXFDebugLogWarning` instead

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -87,17 +92,58 @@ public void RecordRow()
+        {
+            if (currentState != TrackerState.On)
+            {
+                Debug.LogWarning($"Pause command received for tracker in state: '{currentState}'." +
```
Use `Utilities.UXFDebugLogWarning` instead

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -87,17 +92,58 @@ public void RecordRow()
+        {
+            if (currentState != TrackerState.Paused)
+            {
+                Debug.LogWarning($"Resume command received for tracker in state: '{currentState}'." +
```
Use `Utilities.UXFDebugLogWarning` instead

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -115,4 +161,12 @@ public enum TrackerUpdateType
+    /// </summary>
+    public enum TrackerState
+    {
+        On, Off, Paused, Uninitialised
```
I thing Recording, Paused, Stopped are better states. I don't think `Uninitialised` is needed and `Stopped` should be default

jackbrookes commented on Assets/UXF/Scripts/Etc/Tracker.cs:
```diff
@@ -87,17 +92,58 @@ public void RecordRow()
+            {
+                Debug.LogWarning($"Start command received for tracker in state: '{TrackerState.On}'." +
+                    $" This will dump exisiting data! " +
+                    "If you want to restart a paused tracker, use 'ResumeRecording()' instead.");
```
```suggestion
                    $"If you want to restart a paused tracker, use '{nameof(ResumeRecording)}()' instead.");
```

jackbrookes commented:
Looks good other than minor tweaks

martinmmueller commented:
Hey, thanks for the fast reply!
I implemented the changes as you requested, but I changed the calls to the **recording** field in checks (e.g. in **Update()**) to calls to the **Recording** Property (which now wraps the the state check as you suggested) instead of making the explicit check every time. I think it should be more readable this way.
